modules/core/src/main/scala/parser.scala

```diff
@@ -312,71 +316,145 @@ object GraphQLParser {
+      case (t, None) => t
+      case (t, _) => Ast.Type.NonNull(Right(t))
+    }
 
-  lazy val Directives =
+    namedMaybeNull | listMaybeNull
+  }
+
+  lazy val NamedType: Parser[Ast.Type.Named] =
+    Name.map(Ast.Type.Named.apply)
+
+  lazy val Directives: Parser0[List[Ast.Directive]] =
     many(Directive)
 
   lazy val Directive: Parser[Ast.Directive] =
-    keyword("@") ~> (Name, opt(Arguments)).mapN((n, ods) => Ast.Directive(n, ods.orEmpty))
+    keyword("@") *> (Name ~ opt(Arguments)).map { case (n, ods) => Ast.Directive(n, ods.orEmpty)}
 
   lazy val Name: Parser[Ast.Name] = {
     val initial    = ('A' to 'Z').toSet ++ ('a' to 'z').toSet + '_'
     val subsequent = initial ++ ('0' to '9').toSet
-    token((satisfy(initial), many(satisfy(subsequent))).mapN((h, t) => Ast.Name((h :: t).mkString)))
+    token(charIn(initial) ~ many(charIn(subsequent))).map {
+      case (h, t) => Ast.Name((h :: t).mkString)
+    }
   }
 }
 
 object CommentedText {
 
-  def skipWhitespace: Parser[Unit] =
-    takeWhile(c => c.isWhitespace || c == ',').void named "whitespace"
+  lazy val whitespace: Parser[Char] = charWhere(_.isWhitespace)
+
+  def opt[A](parser: Parser0[A]): Parser0[Option[A]] = parser.backtrack.?
+
+  def many[A](parser: Parser[A]): Parser0[List[A]] = parser.rep0
+
+  def skipWhitespace: Parser0[Unit] =
+    charsWhile0(c => c.isWhitespace || c == ',').void.withContext("whitespace")
 
   /** Parser that consumes a comment */
   def comment: Parser[Unit] = {
-    (char('#') ~> many(noneOf("\n\r")) <~ oneOf("\n\r") <~ skipWhitespace).void
+    (char('#') *> many((charWhere(c => c != '\n' && c != '\r'))) <* charIn('\n', '\r') <* skipWhitespace).void.withContext("comment")
   }
 
   /** Turns a parser into one that skips trailing whitespace and comments */
   def token[A](p: Parser[A]): Parser[A] =
-    p <~ skipWhitespace <~ many(comment)
+    p <* skipWhitespace <* many(comment)
+
+  def token0[A](p: Parser0[A]): Parser0[A] =
+    p <* skipWhitespace <* many(comment)
 
   /**
    * Consumes `left` and `right`, including the trailing and preceding whitespace,
    * respectively, and returns the value of `p`.
    */
-  private def _bracket[A,B,C](left: Parser[B], p: => Parser[A], right: => Parser[C]): Parser[A] =
-    token(left) ~> token(p) <~ token(right)
+  private def _bracket[A,B,C](left: Parser[B], p: => Parser0[A], right: => Parser[C]): Parser[A] =
+    token(left) *> token0(p) <* token(right)
 
   /** Turns a parser into one that consumes surrounding parentheses `()` */
-  def parens[A](p: => Parser[A]): Parser[A] =
-    _bracket(char('('), p, char(')')).named(s"parens(${p.toString})")
+  def parens[A](p: => Parser0[A]): Parser[A] =
+    _bracket(char('('), p, char(')')).withContext(s"parens(${p.toString})")
 
   /** Turns a parser into one that consumes surrounding curly braces `{}` */
-  def braces[A](p: => Parser[A]): Parser[A] = {
+  def braces[A](p: => Parser0[A]): Parser[A] = {
```

Comment:
don't need call by name here. I think it is being forced.

If you need laziness, using `Parser.defer` might be more explicit.